### ISSUE TITLE
IRecognition.stop(completion:) completion handler not called

### ISSUE THREAD

**Issue by couchdeveloper** [mediainterface/mispeechkit_4ios#1]
The underlying class conforming to `IRecognition` and created by the SDK, does not call the completion handler in all scenarios.

For example, I created the following extension 

```swift
extension IRecognition {
      
    func stop() async throws {
        try await withCheckedThrowingContinuation { continuation in
            self.stop { ok in
                ok ?
                continuation.resume(with: .success(())) :
                continuation.resume(with: .failure(FailedError()))
            }
        }
    }

```

Wenn running an app, the Swift Runtime reports the following error message in the console:

```console
SWIFT TASK CONTINUATION MISUSE: stop() leaked its continuation!
2022-09-08 15:05:39.350177+0200 MySpeechToText[3234:2166620] SWIFT TASK CONTINUATION MISUSE: stop() leaked its continuation!
```

Obviously, the function `stop(completion:)` has been returned without calling the completion handler.

This probably happens in some error condition, since establishing a session also fails:
Function `IRecognition.initialize(_:completion:)` correctly calls the completion handler, but with a `false` result.
This in turn signals this error:

 **"No vocabularies are available for the user. (id: 1005"** 
 
via the `onError()`  delegate method.

When this fails, I try to tear down the connection calling these `IRecognition` methods :
```
        try await stop()
        try await cleanup()
        try await dispose()
```
(Note, the `IRecognition.cleanup()` and `IRecognition.dispose()` functions which also have completion handlers are modernised to Swift Concurrence style as well, like above)

where `stop()` leaks the continuation, because the completion handler has not been called from within the SDK.

I admit, this might be some edge case. But nonetheless, a completion handler MUST be called eventually and ONCE and only once. ;)

 

**Edit:**

In this use case, when trying to tear down the session using these `IRecognition` methods  (i.e. not calling `stop()`):

```
        try await cleanup()
        try await dispose()
```
 I get the same Runtime Error for `IRecognition.cleanup()`
 
 ```console
 SWIFT TASK CONTINUATION MISUSE: cleanup() leaked its continuation!
```

So, in other words: if the session could not be initialised correctly, it seems we cannot close the session properly due to this issues.
Note that the TCP connection is still open!

**Comment by curthjonas** [comment]
Thank you for creating this issue, we will look into this as soon as possible.
Right now I only can say, unless a session is successfully initialized, there is no need to stop or cleanup a session, because it never have been initialized in the first place.

**Comment by couchdeveloper** [comment]
Thanks for the the quick reply, Jonas.

In our integration, we call a "tear down" (which are these three functions mentioned above), on such an error, in order to completely shut down the connection. In this case, it _seems to me_ a TCP connection is still open when it fails this way. 

I would appreciate it, if these functions just would call its completion handler with "OK" - even they did nothing. This would make it much easier to handle all error scenarios. :)

**Comment by curthjonas** [comment]
With the release of iOS 16 we now plan to discontinue the iOS 14 support and only support iOS 15 and 16. 
Therefore I plan to rework the complete kit and use the async-await-pattern, which was introduced by apple with iOS 15.

**Comment by couchdeveloper** [comment]
> With the release of iOS 16 we now plan to discontinue the iOS 14 support and only support iOS 15 and 16.
> Therefore I plan to rework the complete kit and use the async-await-pattern, which was introduced by apple with iOS 15.

Our plan is to integrate your current version 1.3 into our _legacy_ app which will then require iOS 15 anyway - so I have no gripes with this. :)  

Moving forward is definitely appreciated. Our integration would benefit from API enhancements, too. For example, we would love to see an API, where we get a transcript result which has a flag which says whether the current utterance is completed (true) or whether it is in a "premature" state (false). If this flag is false, the same utterance (as the user proceeds to speak) may be optimised/corrected in the next transcript result by your recognition, until finally, it is considered "completed". In our UI, as long as it is "premature" we _replace_ the previous transcript with the new transcript in our text buffer. Otherwise when the current utterance is considered "completed"  we "commit" it finally in our text buffer and then handle the next result the same way. Optionally - by configuration, the server may send only "completed" transcripts.  

You can support this API, without actually implementing this feature.  ;)

**Comment by curthjonas** [comment]
There already is a feature which sounds like what you want. On initialize you can set the flag "previewResults", with this you can listen to the previewResults-Event, which gives you text considered as not final and may change in the result event.

**Comment by curthjonas** [comment]
There is now a new version v2.0.0 available, which is a complete rewrite of the old miSpeechKit with the new async/await Swift Feature.
